src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelperFactory.cs

```diff
@@ -1,17 +1,37 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Text;
+using Microsoft.Extensions.ObjectPool;
+
 namespace Microsoft.AspNetCore.Mvc.Routing
 {
     /// <summary>
     /// A default implementation of <see cref="IUrlHelperFactory"/>.
     /// </summary>
     public class UrlHelperFactory : IUrlHelperFactory
-    {
-        /// <inheritdoc />
+    {        
         public IUrlHelper GetUrlHelper(ActionContext context)
         {
-            return new UrlHelper(context);
+            var httpContext = context.HttpContext;
+            IUrlHelper urlHelper = null;
+
+            /// Perf: Create only one UrlHelper per context
+            if (httpContext != null && httpContext.Items != null)
```

Comment:
Throw an exception if this is the case. We have a resource already set up for this case: "property X or Y cannot be null".